### PR TITLE
fix(#2281): Prevent double slashes when joining URLs

### PR DESCRIPTION
**Pull request by woubuc**
**What kind of change does this PR introduce?** (check at least one)
- [x] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding Issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)
- [ ] Yes. Issue #___
- [x] No

**The PR fulfills these requirements:**
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: core/tauri/src/manager.rs
Comment (amrbashir) [inline core/tauri/src/manager.rs]: What if `path` doesn't have a leading slash? I think it is better to check for leading slash in `path` instead.
```suggestion
            let separator = match path.starts_with("/") {
```

**Review by amrbashir — CHANGES_REQUESTED** [review]
Thanks for finding and fixing the bug. 
Don't forget to add a change file in `.changes` folder

File: core/tauri/src/manager.rs
Comment (woubuc) [reply core/tauri/src/manager.rs]: In that case I reckon we should check both? Because if either one has a slash, no additional slash should be added in between. Should we also remove one of the slashes if the domain has a trailing slash *and* the path has a leading slash?

Checking the path as well makes this a little more complicated and at that point it might be more advisable to use a proper "url join & normalise" function that handles all sorts of edge cases, rather than rolling our own thing with `format!()` in here.

Comment (amrbashir) [reply core/tauri/src/manager.rs]: yeah that's a valid concern but for the past couple of months we have been adding a slash to `url` no matter what and it worked fine so I don't think there will ever be a trailing slash in `url` but `path` can have a leading slash though. what do you think @lucasfernog ?

Comment (amrbashir) [reply core/tauri/src/manager.rs]: also if you want to make a separate function to handle both cases, go ahead it will be better.

Comment (woubuc) [reply core/tauri/src/manager.rs]: > we have been adding a slash to `url` no matter what and it worked fine so I don't think there will ever be a trailing slash in `url`

I gotta say I'm a little confused by this statement. This PR is a fix for #2281 which is specifically about a trailing slash in `url` that causes it to *not* work fine.

In any case, I'll look into proper URL concatenation and normalisation here because that seems like the better solution anyways. I saw that Tauri already uses the [url](https://crates.io/crates/url) crate internally for the base URLs so I'm gonna explore the codebase a little more to see if we can't just use the [`Url.join()`](https://docs.rs/url/2.2.2/url/struct.Url.html#method.join) function which [does exactly what we need](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=95f4444200e4bfb8e514f931c759c698).

Gonna mark the PR as a draft for now.

Comment (amrbashir) [reply core/tauri/src/manager.rs]: > > we have been adding a slash to `url` no matter what and it worked fine so I don't think there will ever be a trailing slash in `url`
> 
> I gotta say I'm a little confused by this statement. This PR is a fix for #2281 which is specifically about a trailing slash in `url` that causes it to *not* work fine.

Sorry if I wasn't clear, the trailing slash is manually added by us for quite some time now, so you can assume that `url` will never have a trailing slash.

But yeah it is better to check for trailing slash in the base `url` and leading slash in `path` provided by the user

**Comment by woubuc** [comment]
* The custom `format!()` call has been replaced with the [`Url.join()`](https://docs.rs/url/2.2.2/url/struct.Url.html#method.join) function which handles concatenating and normalising URLs a lot better.
  * `WindowManager.get_url()` now returns a `Cow<Url>` to make it possible for other functions to use the `Url` struct directly. I've opted to use `Cow` to prevent cloning the URL since we just need to read it.
  * The relevant tests in `manager.rs` have been updated as well to reflect this change.
  * I've also changed the check for external URLs to compare against `Url.scheme()` instead of converting the entire `Url` to a string first.
* A new `Error` variant `InvalidUrl` is now used to return parse errors that arise from the `url` functions (e.g. when trying to load a path with invalid characters).
* I've also reduced some of the duplicate code in the `WindowManager.get_url()` functions, the cfg-dependant lines are now a separate function and there is only one `match`. I think this makes it a bit clearer what happens here. Do let me know if I've misunderstood what the cfg differences are though.

File: core/tauri/src/manager.rs
Comment (amrbashir) [inline core/tauri/src/manager.rs:204]: just a small nit-pick, the comments are not needed, the code is clear enough

**Review by amrbashir — APPROVED** [review]
LGTM, thanks!

File: core/tauri/src/manager.rs
Comment (woubuc) [reply core/tauri/src/manager.rs:204]: IDE hinting shows function signature and doc comments, not the internal code, so I like to write doc comments for every function no matter how simple it is to make it easier when using them elsewhere in the code. So it's mostly a force of habit. I agree that the code is probably clear enough, but I don't think it's a problem to have these doc comments here?

**Comment by woubuc** [comment]
Looks like I forgot to run Clippy locally before committing.. Should be fixed now.

File: core/tauri/src/manager.rs
Comment (amrbashir) [reply core/tauri/src/manager.rs:204]: It is not a problem, it was just a preference of mine, but I am not the main maintainer anyway so you'd better wait for his review.

**Comment by lemarier** [comment]
Thanks @woubuc !

We do appreciate a lot your contribution!
Have a great day.

**Pull request merged**
